### PR TITLE
Cache favorite names

### LINKED ISSUE #2519: Cache “My Safe” homepage and menu list names
https://www.notion.so/hulking-cosmos-3e8/Q4-2024-QA-Requirements-13e2ee9c5bc280a88458e7a935d17488?pvs=4#13e2ee9c5bc28038961af0c1f1e718a9

and 

https://www.notion.so/hulking-cosmos-3e8/Q4-2024-QA-Requirements-13e2ee9c5bc280a88458e7a935d17488?pvs=4#1422ee9c5bc280bfab30ca28ead3e490

### PR DESCRIPTION
**Pull request by adamgall**
Closes #2519 

Ok let's see.

Modified the structure of the "favorites" object in Local Storage. It used to be an array of strings consisting of `networkPrefix:safeAddress`. Now it's an array of objects consisting of 
```
{
  networkPrefix: string,
  safeAddress: string,
  daoName: string
}
```

So, had to create a migration to convert the old data format into the new data format.

Migrations run _before_ the React app is mounted. So they can't really be async and they need to execute quickly. The migration simply maps the old format to the new format. It then deletes the old Favorites cache in local storage. You'll recognize though, we don't have the DAO name during this migration, so immediately after that migration runs, the `daoName` for each of those favorite items is empty string.

So then, I created a new hook in `Global`, which is responsible for updating the cache favorites names. This is necessary because DAO names aren't forever static! They can be updated!

This hook runs pretty much just once when the app is initially mounted. In our initial case of "the migration just ran", we don't have any Favorite DAO names stored in our local storage cache. So this hook executes, it does the async work necessary to fetch all the Favorites names, and then if any of the newly fetched names are different than the names stored in local storage (which will be _all of them_ initially), it updates those local storage favorites with the newly fetched names and triggers a re-render.

This works well for both our initial case (immediately after the migration runs), and future cases (the DAO has updated their name, and then a user with that DAO favorited goes to our app).

No longer do any components which display Favorites items attempt to fetch the DAO name. Any component that is rendering Favorites data to the screen simply uses the Name which is stored in local storage. It's the above process which is responsible for keeping those names current.

So basically to recap, whenever the app loads, Favorites names are simply read from the local storage. ALSO at the same time, behind the scenes, the app fetches the DAO name from the network, and if there are any changes, it will store those changes into local storage then trigger the app to re-render.

Finally, there were a handful of downstream updates necessary to be able to properly add and remove favorites from local storage, because we have a new piece of data that's being stored (the name).

I think that's it.

### CODE REVIEW

File: src/pages/home/SafeDisplayRow.tsx [hunk 1]
```diff
@@ -24,17 +28,10 @@ export function SafeDisplayRow({ address, network, onClick, showAddress }: SafeM
...
-
-  useEffect(() => {
-    getSafeName(address).then(setSafeName);
-  }, [address, getSafeName]);
-
   const { t } = useTranslation('dashboard');
 
   // if the safe name is an ENS name, let's attempt to get the avatar for that
-  const avatarURL = useAvatar(safeName ?? '');
+  const avatarURL = useAvatar(name ?? '');
```
Comment (DarksightKellar) [inline src/pages/home/SafeDisplayRow.tsx:34]: Should name not be `string` by this point if it's always read from cache and thus should default to empty string if not yet set?

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 2]
```diff
@@ -104,7 +110,7 @@ interface IndexedObject {
 }
 
 export const CACHE_VERSIONS: { [key: string]: number } = Object.freeze({
-  [CacheKeys.FAVORITES]: 1,
+  [CacheKeys.FAVORITES]: 2,
```
Comment (DarksightKellar) [inline src/hooks/utils/cache/cacheDefaults.ts:113]: Hah. Here we are. Full circle. A _part_ of the cache has changed but not the whole, and look -- we can confidently update the version on _just that part_. Nice forward thinking @adamgall 

I wonder if this is a common pattern in cache versioning.

File: src/components/ui/page/Global/index.tsx [hunk 3]
```diff
@@ -17,7 +32,78 @@ const useUserTracking = () => {
...
+
 export function Global() {
   useUserTracking();
-  return <Layout />;
+
+  // Trigger a re-render when favorite names are updated
+  const [favoritesUpdatedTrigger, setFavoritesUpdatedTrigger] = useState(0);
+  useUpdateFavoritesCache(() => setFavoritesUpdatedTrigger(prev => prev + 1));
+
+  return <Layout key={favoritesUpdatedTrigger} />;
```
Comment (DarksightKellar) [inline src/components/ui/page/Global/index.tsx:115]: Just noting here, that ideally it'd be nice to have each favorite-showing component (or the whole list) directly listen to and re-render if the names are updated instead of the whole app re-rendering.

Topic for another day though.

File: src/components/ui/menus/SafesMenu/SafeMenuItem.tsx [hunk 4]
```diff
@@ -30,15 +29,8 @@ export function SafeMenuItem({ address, network }: SafeMenuItemProps) {
...
-  const { getSafeName } = useGetSafeName(getChainIdFromPrefix(network));
-  const [safeName, setSafeName] = useState<string>();
-
-  useEffect(() => {
-    getSafeName(address).then(setSafeName);
-  }, [address, getSafeName]);
-
   // if by chance the safe name is an ENS name, let's attempt to get the avatar for that
-  const avatarURL = useAvatar(safeName ?? '');
+  const avatarURL = useAvatar(name ?? '');
```
Comment (DarksightKellar) [inline src/components/ui/menus/SafesMenu/SafeMenuItem.tsx:33]: Similar question here on `name` being `undefined`-able. Beginning to look like this was on purpose. Loading indicator?

File: src/components/ui/menus/DAOSearch/SearchDisplay.tsx [hunk 5]
```diff
@@ -90,6 +90,7 @@ export function SearchDisplay({
             {t(isCurrentSafe ? 'labelCurrentDAO' : 'labelDAOFound')}
           </Text>
           <SafeDisplayRow
+            name={undefined}
```
[the remark this answers is not in the document]
Comment (DarksightKellar) [reply src/components/ui/menus/DAOSearch/SearchDisplay.tsx:93]: These are search results and rather likely not something the user will have saved

[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/menus/DAOSearch/SearchDisplay.tsx:93]: The `favorites` cache?

File: src/components/ui/page/Global/index.tsx [hunk 6]
```diff
@@ -17,7 +32,78 @@ const useUserTracking = () => {
...
+        const allFavorites = favoritesList.map(original => {
+          const updated = updatedFavorites.find(
+            update =>
+              update.networkPrefix === original.networkPrefix &&
+              update.address === original.address,
+          );
+          return updated || original;
+        });
+
+        setValue({ cacheName: CacheKeys.FAVORITES }, allFavorites, CacheExpiry.NEVER);
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/page/Global/index.tsx:94]: I think you're missing something... the code in this Global/index.tsx specifically exists to update the favorites list with the new name if that name has changed. We're talking about _favorites_ here, we don't want this "cache" (it's really not a cache) to ever expire.

File: src/hooks/utils/useGetSafeName.ts [hunk 7]
```diff
@@ -1,54 +1,60 @@
...
+  publicClient: PublicClient | undefined,
+  subgraph: { space: number; slug: string; version: string },
+  address: Address,
+) => {
+  if (!publicClient) {
+    throw new Error('Public client not available');
+  }
+
+  const ensName = await publicClient.getEnsName({ address }).catch((error: Error) => {
+    if (error.name === 'ChainDoesNotSupportContract') {
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/hooks/utils/useGetSafeName.ts:19]: Not sure. I didn't modify this code, it's been around for a while.

File: src/components/ui/page/Global/index.tsx [hunk 8]
```diff
@@ -17,7 +32,78 @@ const useUserTracking = () => {
...
+          }
+
+          const favoriteNetwork = supportedNetworks.find(
+            network => network.chain.id === favoriteChain.id,
+          );
+          if (!favoriteNetwork) {
+            return;
+          }
+
+          const favoritePublicClient = createPublicClient({
```
[the remark this answers is not in the document]
Comment (DarksightKellar) [reply src/components/ui/page/Global/index.tsx]: Yeah I'm a fan of not re-initialising when already previously done. Maybe keep a running map of networks -> client in this hook, pre-filled with the currently connected network client?

[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/page/Global/index.tsx]: pushed

File: src/hooks/utils/useGetSafeName.ts [hunk 9]
```diff
@@ -1,54 +1,60 @@
...
 import { useNetworkConfig } from '../../providers/NetworkConfig/NetworkConfigProvider';
 import { createAccountSubstring } from './useGetAccountName';
 
+export const getSafeName = async (
+  publicClient: PublicClient | undefined,
+  subgraph: { space: number; slug: string; version: string },
+  address: Address,
+) => {
+  if (!publicClient) {
+    throw new Error('Public client not available');
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/hooks/utils/useGetSafeName.ts:15]: yeah good callout. let me work on this.

**Review by DarksightKellar — CHANGES_REQUESTED** [review]
One or two requests, I think, then looks good to me.

File: src/hooks/utils/useGetSafeName.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (adamgall) [reply src/hooks/utils/useGetSafeName.ts]: Check my latest commit. I move where the error is thrown.

File: src/components/ui/menus/SafesMenu/SafeMenuItem.tsx (hunk 4 above)
Comment (adamgall) [reply src/components/ui/menus/SafesMenu/SafeMenuItem.tsx]: pushed

File: src/pages/home/SafeDisplayRow.tsx (hunk 1 above)
Comment (adamgall) [reply src/pages/home/SafeDisplayRow.tsx:42]: Thanks for calling this out. See: https://github.com/decentdao/decent-interface/pull/2559#discussion_r1852154882

File: src/components/ui/page/Global/index.tsx (hunk 3 above)
Comment (adamgall) [reply src/components/ui/page/Global/index.tsx:115]: Yes agreed! That's what I was struggling with which led me to write this code.

**Review by DarksightKellar — APPROVED** [review]

File: src/components/ui/menus/DAOSearch/SearchDisplay.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (mudrila) [reply src/components/ui/menus/DAOSearch/SearchDisplay.tsx:93]: Why not giving it a chance though? 
Could be over-engineering, not holding this opinion strong

File: src/components/ui/page/Global/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (mudrila) [reply src/components/ui/page/Global/index.tsx:101]: Oh, yeah - you're right - we don't want whole list of favorites to expire ever. However, names within these favorites might expire. I seem to miss how we catch the fact that name was changed so gonna trust you it's there 🤣

File: src/components/ui/menus/DAOSearch/SearchDisplay.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/menus/DAOSearch/SearchDisplay.tsx:93]: yeah i think this would be over-engineering. to @parkermccurley 's point, search is rarely used in practice, and trying to shoehorn our favorites data structure into search results seems like the kind of code bloat that i'm trying to cut down on.

**Review by mudrila — APPROVED** [review]

File: src/components/ui/page/Global/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/page/Global/index.tsx:101]: Yeah it's in here, it's basically what the entirety of the useEffect in this file is for!

On app load, we grab all the favorites from local storage, go through each one and look up its _current_ name, then overwrite the name in local storage if it's different from what we already have saved.

**Comment by adamgall** [comment]
Before I merge, @mudrila @DarksightKellar @Da-Colon will you please pull down this code locally and and run it and make sure your "favorites" on localhost:3000/:8888 are properly migrated and this works on your machine?

File: src/hooks/utils/useGetSafeName.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/hooks/utils/useGetSafeName.ts:15]: ITs called `ChainDoesNotSupportContractErrorType`

**Review by Da-Colon — APPROVED** [review]
Migrations looks successful.

File: src/hooks/utils/useGetSafeName.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/hooks/utils/useGetSafeName.ts:15]: `.name` will give you what you want @adamgall. You COULD add it. but everything works

[the remark this answers is not in the document]
Comment (adamgall) [reply src/hooks/utils/useGetSafeName.ts]: I dig it. Pushed.
